security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,375 @@
+	ServerNameOverride string
+}
+
+// ServerOptions contains all the fields and functions needed to be filled by the client.
+type ServerOptions struct {
+	// ---------------General Rules For Certificate Setting-----------------------------------------
+	// Certificates or GetClientCertificate indicates the certificates sent from the server to
+	// the client to prove server's identities. The rules for setting these two fields are:
+	// Either Certificates or GetCertificate must be set; the other will be ignored
+	//----------------------------------------------------------------------------------------------
+	// If field Certificates is set, field GetClientCertificate will be ignored. The server will use
+	// Certificates every time when asked for a certificate, without performing certificate reloading.
+	Certificates []tls.Certificate
+	// If GetClientCertificate is set and Certificates is nil, the server will invoke this
+	// function every time asked to present certificates to the client when a new connection is
+	// established. This is known as peer certificate reloading.
+	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	// ---------------General Rules For Root Certificate Setting------------------------------------
+	// RootCACerts or GetRootCAs indicates the the certificates trusted by the server side. The rules
+	// for setting these two fields are:
+	// If requiring mutual authentication on server side:
+	//     Either RootCACerts or GetRootCAs must be set; the other will be ignored
+	// Otherwise:
+	//     Nothing needed(the two fields will be ignored)
+	//----------------------------------------------------------------------------------------------
+	// If field RootCACerts is set, field GetRootCAs will be ignored. The server will use RootCACerts
+	// every time when verifying the peer certificates, without performing root certificate reloading.
+	RootCACerts *x509.CertPool
+	// If GetRootCAs is set and RootCACerts is nil, the server will invoke this function every time
+	// asked to check certificates sent from the client when a new connection is established.
+	// This is known as trust certificate reloading.
+	GetRootCAs func(rawConn net.Conn, rawCerts [][]byte) (*x509.CertPool, error)
+	// If the server want the client to send certificates to prove its identity.
+	MutualAuth bool
+}
+
+func (o *ClientOptions) config() (*tls.Config, error) {
+	if o.RootCACerts == nil && o.GetRootCAs == nil {
+		return nil, fmt.Errorf("either RootCACerts or GetRootCAs must be specified")
+	}
+	config := &tls.Config{
+		ServerName:           o.ServerNameOverride,
+		Certificates:         o.Certificates,
+		GetClientCertificate: o.GetClientCertificate,
+		RootCAs:              o.RootCACerts,
+	}
+	// We fall back to normal config settings if users don't need any of the advanced TLS features.
+	if o.GetClientCertificate == nil && o.GetRootCAs == nil && o.VerifyPeer == nil {
```

Comment:
Why not always set InsecureSkipVerify = true?
In any case, VerifyPeerCertificate from buildVerifyFunc will be called.

Reply:
Done.